### PR TITLE
Implementation of output-unused-css

### PR DESCRIPTION
**Pull request by kevinramharak**
## Proposed changes

This PR implements an option to output the purged css alongside the regular output. This is a feature request mentioned in #131.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

<details>
  <summary>
    Old comments
  </summary>

- ~~a few tests are failing because of Windows files paths. The tests fail because they check for `/` seperators instead of `\\`~~.
- ~~other tests are performance related. Not sure what is causing them to fail sometimes. Maybe my laptop is just a bit to slow to hit the 5 second threshold everytime? These changes should only have a performace penalty if the option is actually being used, but since they are off by default I can't see how they could.~~
- ~~some tests are failing on these type of lines, I really don't get how these changes could have impacted those.~~

```
+
    - /*!************************************************************************!*\
    + /*!************************************************************************!*\
    -   !*** css ../../../node_modules/css-loader/dist/cjs.js!./src/style.css ***!
    +   !*** css ../../../node_modules/css-loader/dist/cjs.js!./src/style.css ***!
    -   \************************************************************************/
    +   \************************************************************************/
    
```
- ~~the last test that is failing seems to choke on a `JSON.parse` line. I'm checking to see if these changes have anything to do with it.~~
```
  ● PurgeCSS CLI › should print the correct output

    SyntaxError: Unexpected end of JSON input
        at JSON.parse (<anonymous>)

      13 |     );
      14 |     const result = JSON.parse(response.stdout);
    > 15 |     expect(result[0].css).toBe(".hello {\n  color: red;\n}\n");
         |                             ^
      16 |   });
      17 | });
      18 |
```
</details>

I'm looking for feedback on:
- What the CLI and programmatic API changes should look like. Currently its  `rejectedCss` and `--rejected-css`
- What the tests should look like, I'm not sure what the edge cases are that tests should cover in regard to this feature.
- ~~The performance hit that will occur when enabling this feature.~~

### CODE REVIEW

**Comment by kevinramharak** [comment]
Seems like the Github CI tests are passing. So I'm going to assume the failing tests are caused by running on a windows machine.

**Comment by kevinramharak** [comment]
I did a test on a project im working on:
```
# ls -l dist/*.css
105086 - dist/style.critical.css
606847 - dist/style.css
492651 - dist/style.rejected.css
```
These added up:
```
all               == 606_847
critical + rejected
105_086 + 492_651 == 597_737
```
So about `9000` bytes seem to be missing, but that seems pretty close.

**Comment by kevinramharak** [comment]
@Ffloriel Can this get a review? I added tests and updated the documentation. I would like to discuss the API, plugin integrations and if more tests are needed.

File: packages/purgecss/src/index.ts [hunk 1]
```diff
@@ -482,7 +485,13 @@ class PurgeCSS {
...
     const parent = node.parent;
-    if (!node.selector) node.remove();
+    if (!node.selector) {
+      node.remove()
+      if (this.options.rejectedCss) {
+        node.selector = originalSelector;
+        this.removedNodes.push(node);
+      }
+    }
     if (isRuleEmpty(parent)) parent?.remove();
```
Comment (Ffloriel) [inline packages/purgecss/src/index.ts:495]: There is the situation when you have an empty parent at the result of the node being removed. In that situation, I think you might want to add the parent node to the `removedNodes` array.

For:
```
@media (max-width: 66666px){
  .unused-class, .unused-class2 {
    color: black;
  }
}
```
We probably want to ouput the same for the rejectedCss, but it currently outputs:
```
rejectedCss: '\n  .unused-class, .unused-class2 {\n    color: black;\n  }'
```

**Review by Ffloriel — CHANGES_REQUESTED** [review]
Thanks for taking the time to take care of the most requested feature!
There's one little comment about the parent node not being in the rejected CSS but apart from that, LGTM!

File: packages/purgecss/src/index.ts (hunk 1 above)
Comment (kevinramharak) [reply packages/purgecss/src/index.ts:495]: I'll add that test case and see how I can resolve it.

**Review by Ffloriel — APPROVED** [review]

File: packages/purgecss/src/index.ts (hunk 1 above)
Comment (kevinramharak) [reply packages/purgecss/src/index.ts]: @Ffloriel Thank you for your comment, it helped me understand a little more about how the code works. I added the following test case:
```css
@media (max-width: 66666px) {
  .used-class, .unused-class {
    color: black;
  }
}
```
Which I was handling incorrectly.
What the rejected css 'algorithm' now does is:
- keep track of what selectors are removed from the rule
- if any selector is removed from the rule, clone the rule (or its parent if there exists one)
- replace the (deemed critical) selectors from the clone with the rejected selectors
- preserve that rule node as `rejectedCss` output

So the above case would result in:
```css
/* critical */
@media (max-width: 66666px) {
  .used-class {
    color: black;
  }
}
/* rejected */
@media (max-width: 66666px) {
  .unused-class {
    color: black;
  }
}
```

Purging the css is easier since you walk down the AST nodes and remove what you don't need. Keeping the rejected css is harder since once you have a node you need to preserve you then also have to preserve its ancestors to get the correct output.

Can you think of any examples where a rule node is nested more that just 1 parent deep? My understanding is that these are CSS nodes thus the selectors are flattened, but if there is a case like:
```css
@media (min-width: 1px) {
  @media (max-width: 200px) {
    .selector {
      color: black;
    }
  }
}
```
I would have to add something that goes up the tree to preserve the correcct output

**Comment by kevinramharak** [comment]
@Ffloriel Wups I see you already merged it but I found a flaw in my output generation. I was about to push a commit that fixes it. Thanks for the fast response tough

File: packages/purgecss/src/index.ts (hunk 1 above)
Comment (kevinramharak) [reply packages/purgecss/src/index.ts]: Added a new PR that fixes the above [here](https://github.com/FullHuman/purgecss/pull/799)
